subgraphs/_Metrics/BlockSpeedForecast.ts

```diff
@@ -0,0 +1,142 @@
+// 86400 = 1 Day
+export const RATE_IN_SECONDS = 86400
+export const RATE_IN_SECONDS_BD = BigDecimal.fromString(RATE_IN_SECONDS.toString())
+
+// WINDOW_SIZE_SECONDS and BLOCK_STORAGE_INTERVALS can be modified. These are just recommended values - 'somewhat' arbitrary. 
+
+// The window size determines the range of blocks that you track from the current block minus the window size.
+// Window of block time used to calculate the moving average.
+// Increasing means less deviation but also less sensitivity to changing block speeds.
+export const WINDOW_SIZE_SECONDS = 86400
+export const WINDOW_SIZE_SECONDS_BD = BigDecimal.fromString(WINDOW_SIZE_SECONDS.toString())
+
+// The storage interval tells you to only store blocks that are space by at least this amount. 
+// Increasing this value will mean less blocks stored and less frequently computes blocksPerDay.
+export const BLOCK_STORAGE_INTERVAL = 5 as i32
+
+// Paramaters: [event.block.timestamp, event.block.number]
+// Call this function in event handlers frequently enough so that it updates on most blocks 
+export function updateMovingAverageBlockSpeedForecast(event: ethereum): void {
+
+    let blockTracker = getOrCreateBlockTracker()
+    let blocks: string[]
+    blocks = blockTracker.blocks
+
+    // Create entity for the current block
+    let currentTimestamp = event.block.timestamp.toI32()
+    let currentBlockNumber = event.block.number.toI32()
+    let currentBlock = getOrCreateBlock(currentTimestamp, currentBlockNumber)
+
+    if (blocks.length == INT_ZERO) {
+        blocks.push(currentBlock.id)
+        blockTracker.blocks = blocks
+
+        blockTracker.save()
+        currentBlock.save()
+
+        // return because there is only 1 reference point.
+        return
+    }
+
+    // Add current timestamp and block numnber to array if new block is at least X blocks later than previously stored.
+    // Used to save memory and efficiency on array resizing.
+    let recentSavedBlock = _Block.load(blocks[blocks.length - INT_ONE])
+    if (currentBlockNumber - recentSavedBlock!.blockNumber > BLOCK_STORAGE_INTERVAL) {
+        blocks.push(currentBlock.id)
+        currentBlock.save()
+    }
+
+    else return
+
+    // Return 0 incase there is only one element.
+    if (blocks.length < INT_TWO) {
+        return
+    }
+
+    // The timestamp at the start of the window (default 24 hours in seconds).
+    let startTimestamp = currentTimestamp - WINDOW_SIZE_SECONDS
+
+    // Make sure you still have 2 blocks to calculate rate (This shouldn't really happen past the beginning).
+    while(blocks.length > INT_TWO) {
+        let oldestPushedBlock = _Block.load(blocks[INT_ZERO])
+
+        // Remove items from array and storage that were greater than 24 Hours ago.
+        if (oldestPushedBlock!.timestamp < startTimestamp) {
+            store.remove('_Block', oldestPushedBlock!.id)
+            blocks.shift()
```

Comment:
This is a little bit inefficient: https://stackoverflow.com/questions/22614237/javascript-runtime-complexity-of-array-functions

Reply:
I understand that it is a little inefficient. There is not a native function for slicing off the first index while still getting O(1) pushing to the back. Would you suggest using a pointer to track the window and only slicing off the front of the array once a day or something?

Reply:
yeah, you can use a circular buffer which will give you constant time